File: rust-example/src/capabilities/book_stay/context.rs
use chrono::NaiveDate;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GuestBookingEligibility {
    Eligible,
    Blocked,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ListingBookingStatus {
    Bookable,
    Disabled,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ListingBookingSettings {
    pub booking_status: ListingBookingStatus,
    pub max_guests: u32,
    pub min_nights: u32,
    pub max_nights: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BookingContext {
    pub guest: Option<GuestBookingEligibility>,
    pub listing: Option<ListingBookingSettings>,
    pub unavailable_nights: Vec<NaiveDate>,
}

File: csharp-example/src/BookingAStay/Capabilities/BookStay/Context.cs
namespace BookingAStay.Capabilities.BookStay;

public enum GuestBookingEligibility
{
    Eligible,
    Blocked,
}

public enum ListingBookingStatus
{
    Bookable,
    Disabled,
}

public sealed record ListingBookingSettings(
    ListingBookingStatus BookingStatus,
    int MaxGuests,
    int MinNights,
    int? MaxNights);

// Input value for the BookStay Functional Core.
public sealed record BookingContext(
    GuestBookingEligibility? Guest,
    ListingBookingSettings? Listing,
    IReadOnlyList<DateOnly> UnavailableNights);
